File: DSA/Stack/Code/PopOperation.py
# Pop Operation in Python

class Stack:
    def __init__(self, max_size=100):
        self.arr = [0] * max_size
        self.top = -1
        self.max_size = max_size
    
    def is_empty(self):
        return self.top == -1
    
    def push(self, value):
        self.top += 1
        self.arr[self.top] = value
    
    def pop(self):
        if self.is_empty():
            print("Stack underflow - cannot pop")
            return -1
        value = self.arr[self.top]
        print(f"Successfully popped: {value} (top = {self.top - 1})")
        self.top -= 1
        return value
    
    def display(self):
        if self.is_empty():
            print("Stack is empty")
            return
        print("Current stack: ", end="")
        for i in range(self.top + 1):
            print(self.arr[i], end=" ")
        print()

# Test
stack = Stack()

print("=== Pop Operation ===\n")

for i in range(1, 6):
    stack.push(i * 10)

stack.display()

print("\nPopping elements:")
stack.pop()
stack.pop()

stack.display()

print("\nPopping all remaining:")
while not stack.is_empty():
    stack.pop()

stack.display()
stack.pop()

print("\nComplexity Analysis:")
print("Time: O(1)")
print("Space: O(1)")

File: DSA/Stack/Code/PopOperation.cpp
#include <iostream>
#include <vector>
using namespace std;

// Pop Operation in C++
class Stack {
private:
    vector<int> arr;
    int top;
    
public:
    Stack() : top(-1) { arr.resize(100); }
    
    bool isEmpty() {
        return top == -1;
    }
    
    void push(int value) {
        arr[++top] = value;
    }
    
    int pop() {
        if (isEmpty()) {
            cout << "Stack underflow - cannot pop\n";
            return -1;
        }
        int value = arr[top];
        cout << "Successfully popped: " << value << " (top = " << top - 1 << ")\n";
        top--;
        return value;
    }
    
    void display() {
        if (isEmpty()) {
            cout << "Stack is empty\n";
            return;
        }
        cout << "Current stack: ";
        for (int i = 0; i <= top; i++) {
            cout << arr[i] << " ";
        }
        cout << "\n";
    }
};

int main() {
    Stack stack;
    
    cout << "=== Pop Operation ===\n\n";
    
    for (int i = 1; i <= 5; i++) {
        stack.push(i * 10);
    }
    
    stack.display();
    
    cout << "\nPopping elements:\n";
    stack.pop();
    stack.pop();
    
    stack.display();
    
    cout << "\nPopping all remaining:\n";
    while (!stack.isEmpty()) {
        stack.pop();
    }
    
    stack.display();
    stack.pop();
    
    cout << "\nComplexity Analysis:\n";
    cout << "Time: O(1)\n";
    cout << "Space: O(1)\n";
    
    return 0;
}
